### PR TITLE
feat(network): permanent peer store

### LINKED ISSUE #866: Keeping peer list when node goes off
We need to keep peers somewhere when node goes off, we can use them when node try to start and sync again.

### PR DESCRIPTION
**Pull request by kehiy**
## Description

This PR adds a permanent json-based peer store. It keeps a backup of all connected peers at runtime when we turn the node off and restore/use it when we try to start our node again.

This change will enhance the syncing process. here is a part of `peers.json` after my test for a few second:

```json
[
    "/ip4/5.161.222.154/tcp/21888/p2p/12D3KooWQN2VN1RWQ5eWq8rVexpBnwVSy7k34Rpep4R6V57XWYau",
    "/ip4/135.181.97.206/tcp/21888/p2p/12D3KooWBC5YQJdRAvupSXUPNYHiSctzZeTPNzen1JsRwj9aEyYZ",
    "/ip4/65.21.69.53/tcp/21888/p2p/12D3KooWQ1izUE5jcmMyjmxRdb2CmASe8Nfn81UL2G5rYjptC2RV",
    "/ip4/45.134.226.48/tcp/21888/p2p/12D3KooWBSP8AdiHkkvPunpbhfZE96YTY6om4hw35Xpe3bJFBUjh",
    "/ip4/104.234.1.82/tcp/21888/p2p/12D3KooWK1z7QAskVrQd98r98UMSPwTLr4out8B9NkQTrCdZPCZx",
    "/dns/pactus-bootnode.stake.works/tcp/21888/p2p/12D3KooWJPuhW7Ejmb8KprpVNRzn1eUQXDHU3n5eTo2vodgvVxoJ",
    "/ip4/37.27.26.40/tcp/21888/p2p/12D3KooWPehhWgfdZG6x9zE7QdQxuAcfbQPAyBPjcXLfVYfU9FAd",
    "/ip4/185.216.75.142/tcp/21888/p2p/12D3KooWNjYFZpq5YwdUC6JgHfkZhUY7SyxTAQCHrdNwf2t1FXSn",
    "/ip4/84.247.172.23/tcp/21888/p2p/12D3KooWFJYk55mUD6FR9Gx2YGQ34PYziwTQhY1kZ3TMS6fGUz6g",
    "/ip4/84.247.165.249/tcp/21888/p2p/12D3KooWQmv2FcNQfh1EhA98twt8ePdkQaxEPeYfinEYyVS16juY",
    "/dns/pactus-bootstrap.thd.io.vn/tcp/21888/p2p/12D3KooWDgsSSGJzihBJxqNJFqbGwbTEDXRoUBph1BCfwm11R6KY",
    "/ip4/159.148.146.149/tcp/21888/p2p/12D3KooWKCokWtpdudxgsLRoFcnrW35vhn6w632iGWCga7E5e68Q",
    "/ip4/84.247.184.95/tcp/21888/p2p/12D3KooWSjp5GFmCCsYkKcqicrghyHqfjdDCfLfgfqWmG5JTQSDu",
    "/ip4/161.97.145.100/tcp/21888/p2p/12D3KooWCjZ35Hu8Mt5GjZ6fJzJBZuUSsAppAzXcbMcXkswR4HrU",
    "/dns/pactus-bootstrap.teoviteovi.com/tcp/21888/p2p/12D3KooWMP5qWNHF1RfzhpAkBmQZzBfbaSJZCo3QF1fxs5CFXkv8",
    "/ip4/195.201.112.164/tcp/21888/p2p/12D3KooWKiYGT4mCapgcqLPSwbkJY1KgZMW5hdC8XJMQSU23tiSL",
    "/ip4/37.60.245.47/tcp/21888/p2p/12D3KooWBH27VEPdGBYzq1vYciRgnDn3ozStRaUYd2vJWGgaKEUZ",
    "/ip4/128.140.103.90/tcp/21888/p2p/12D3KooWN4eypkp61j12SMSkmrEfQv4RjnggmRfcEfom4mcKUpsR",
    "/dns/pactus-bootstrap.sensifai.com/tcp/21888/p2p/12D3KooWLpgBQt7A1W8FhTLVYwQMEnTBEX7uudLvd1YMZsDHqqWh",
    "/ip4/75.119.159.113/tcp/21888/p2p/12D3KooWL7nUsmd7JNrhrfbBjg6JxPFiZA5LXbdCDPnw5Xpdp4a9",
    "/ip4/173.212.254.120/tcp/21888/p2p/12D3KooWJR19vjuBzuJTceCoy3LG8jggkHCBGb9j5SzdKtoPL7Yb",
    "/ip4/194.163.164.10/tcp/21888/p2p/12D3KooWDRFCW5ksbnn6Uq7kx3HZt3EnrKjujqqs8QBQmA7TK5o6",
    "/ip4/62.171.134.3/tcp/21888/p2p/12D3KooWSv5jsXMg1b4PBd2SCcPLj3Mc5KzzBYnpPkiS7JjuqhHF",
    "/ip4/94.72.120.50/tcp/21888/p2p/12D3KooWCJX9aLnzmEFtEasvbbaFqHLTboFVFptZxeFnbnSxpb2m",
    "/ip4/65.21.180.80/tcp/21888/p2p/12D3KooWANqay6T1xCrK3imfAEr3GpiGzxT6SWCEsRapo8tJWAZN",
    "/ip4/42.117.247.180/tcp/21888/p2p/12D3KooWLVZSkgkFgrRNMfkCtULqtErYcNgjqgueeH944vDpDCPv",
    "/ip4/158.220.91.129/tcp/21888/p2p/12D3KooWGfehwyZJBu6WUMLyzJ8sDJachvhxhqw5Wh4uU3H99GLq",
    "/ip4/65.21.155.128/tcp/21888/p2p/12D3KooWPykSnzWpEasRijCgQ99vxAxWKL32zY24MmvtaXsHapSf",
    "/ip4/20.174.33.162/tcp/21888/p2p/12D3KooWKg6aLa77yAaqMCb45aH5iQuTr5GzRUWUCJ1sZYB5vnoL",
    "/ip4/43.135.166.197/tcp/21888/p2p/12D3KooWEqyjMNzFe92HFZMZQd3YZ6zvztuPaC3a4wAUvPf4moyP",
    "/ip4/109.123.253.88/tcp/21888/p2p/12D3KooWFUJhDjYcAXgEnW1pTLRDWnH9CLgBPcezgU7pCpxMngt8",
    "/dns/bootstrap1.pactus.org/tcp/21888/p2p/12D3KooWMnDsu8TDTk2VV8uD8zsNSB6eUkqtQs6ttg4bHq9zNaBe",
    "/ip4/37.60.241.89/tcp/21888/p2p/12D3KooWMZtkk5HS69ukYYdfxD3CgBgTtKZg7kv43mCz1HMrDnQw",
    "/ip4/37.60.233.235/tcp/21888/p2p/12D3KooWNzbuBESYzbKGm7UyLkD22cHSgjRMUx4rjEKLBfVSZuY1",
    "/dns/bootstrap2.pactus.org/tcp/21888/p2p/12D3KooWM39ag7ghta49qybf7McADgT8FLakTYkCsiBvwdnjuG5q",
    "/ip4/157.90.22.45/tcp/21888/p2p/12D3KooWLUYJ55DyJjSZvwPJDQNVYSBYqiMxJ9bFzRcVBgvSkN9L",
    "/dns/bootstrap3.pactus.org/tcp/21888/p2p/12D3KooWBCPSZWheet6tMoHbVBCDfBwQm4yzCwcQ8hJ6NMCN97sj",
    "/ip4/167.86.121.11/tcp/21888/p2p/12D3KooWMpVXMQkr26ff3y8gC5z1h9RZNxTvW4WzZBML8GHHUoL5",
    "/dns/pactus-bootstrap.stakers.world/tcp/21888/p2p/12D3KooWK1DAcxf8RJA2PVkwr9ehk1zgXcnjLyPCsYCpvw1NSfhc",
    "/ip4/154.38.187.40/tcp/21888/p2p/12D3KooWQyt7JCyhhnZ8jyT1vso4GGYVbpJhoXJy5KRpnqa8DmPg",
    "/dns/pactus-bootstrap.codeblocklabs.com/tcp/21888/p2p/12D3KooWSdmPLqWbmv5x5pBiL5nHwKN5S2jhP96oXCKmoEmAhV4Z",
    "/ip4/46.250.233.5/tcp/21888/p2p/12D3KooWH1kam9g1QFWmtWMLzP2MDY4ptpLTMPU8wSq3Q4k3rvnn",
    "/ip4/49.13.27.37/tcp/21888/p2p/12D3KooWPSh9KMa6tRfSewK3sUC4BDSPnXseVDGJxWqNZxKcXnGM",
    "/ip4/65.108.68.214/tcp/21888/p2p/12D3KooWD7X8BDqV9r6zGmoo2dpMYo4Z6YoyDQp6j4UnB2Bj3Fq2",
    "/ip4/62.171.130.196/tcp/21888/p2p/12D3KooWSkwB8AiWz3VGyG5cgC3ry5dmCw3TthQNc1jvRNe9EEKH",
    "/dns/pactus-bootstrap.ionode.online/tcp/21888/p2p/12D3KooWG5tvvfS8PRd2Bgyxxqrok7QfohkHSprFPxmT3HTh8dVD",
    "/ip4/85.190.246.251/tcp/21888/p2p/12D3KooWMTPKpr6EHMmZ2NSXkup4NaqwKnYmScHEsN3HpcqYgBZL",
    "/ip4/116.203.238.65/tcp/21888/p2p/12D3KooWKf32k9dLE4fA5tkh4KvvEAQDmHxHLWiahjpt2Hfx3d88",
    "/ip4/135.181.42.222/tcp/21888/p2p/12D3KooWN46UYQNoZvzd9GiJPQC7LByj6RAg8rRxdUTiQ24vHsN1",
    "/ip4/65.109.143.78/tcp/21888/p2p/12D3KooWPnWgEFV32cvpfuv98oioZQJJKvRM46BEsFaoXYijvgvc",
    "/ip4/95.111.242.225/tcp/21888/p2p/12D3KooWRNTZpo1XNWKQx2Z5HEu5RnqKeNDV5y4qMKX5f4aLtbv9",
    "/dns/pactus-bootstrap.mflow.tech/tcp/21888/p2p/12D3KooWJD2PBAur6kXjX31dyrPzjSV947ppZ885QDKCXEvLm9Vb",
    "/ip4/95.216.222.135/tcp/21888/p2p/12D3KooWHicbZ3dBnR7JaVKr4U7WNUWyhDnUn2TwMVmPqW8Ftqop",
    "/ip4/46.196.214.35/tcp/21888/p2p/12D3KooWJ3H7YAyemqxtToG6yRRzTfLdZD3eJk2xsQRaKxfdNhKf",
    "/ip4/109.123.240.5/tcp/21888/p2p/12D3KooWCT92yaS4JgrHHzeRPsJxaLzRdSEknGEN4HxQFfscoY8k",
    "/ip4/167.86.116.61/tcp/21888/p2p/12D3KooWAmc69XaFh6T76BQQaFmSakYFgDyvzDLfVBgTzqmNtyXM",
    "/ip4/65.108.211.187/tcp/21888/p2p/12D3KooWGLrSiE5KWoc9GpXfryvKqheHj6j78ryMDDC17SY4YoCi",
    "/ip4/167.86.80.169/tcp/21888/p2p/12D3KooWDqAPjeRR41GYoQw9RQaSZdFx59k3SpkR29yAycpJUhfF",
    "/ip4/38.242.225.172/tcp/21888/p2p/12D3KooWDTP7CKGvbcg5WiuKB2Bt4acf2LWMdbUjMqU63dKKCVMB",
    "/ip4/37.27.25.245/tcp/21888/p2p/12D3KooWAmhn86b5igYjzFbaJddE57xT9UpZePFiEoc3jcuYHDvn",
    "/ip4/194.163.189.255/tcp/21888/p2p/12D3KooWExbN79BrgDWjdLUd8TZaP2LvCDN8zJkMuPhRHkjKDcJS",
    "/ip4/65.21.152.153/tcp/21888/p2p/12D3KooWCirKBMmhmL6a4Ng5TVHQ4Leb6Wfj4M8s9raUzdFQn5aP"
]
```

I suggest you guys test it on your side with your environment to ensure it works properly.

## Related issue

- Fixes #866

### CODE REVIEW

**Comment by kehiy** [comment]
Tree additional notes:

1. The `peers.json` can be shared offline same as offline chain data download for better syncing. (for example, I'm running a new node so I can use the previous node peer store to have more seed nodes to connect to.)

2. We can add a disable option for this peer store in the config but I thought it was unnecessary.

3. offline nodes, private IPs, etc will be ignored, and checking them or removing them from the peer manager and saving the store then is not very efficient or helpful. (simply won't make anything worse or better.)

**Review by Ja7ad — CHANGES_REQUESTED** [review]

File: network/peermgr.go [hunk 1]
```diff
@@ -2,16 +2,21 @@ package network
...
 
 	lp2phost "github.com/libp2p/go-libp2p/core/host"
 	lp2pnet "github.com/libp2p/go-libp2p/core/network"
 	lp2ppeer "github.com/libp2p/go-libp2p/core/peer"
 	"github.com/multiformats/go-multiaddr"
+	"github.com/pactus-project/pactus/util"
 	"github.com/pactus-project/pactus/util/logger"
 )
 
+const PeerStorePath = "peers.json"
```
[the remark this answers is not in the document]
Comment (kehiy) [reply network/peermgr.go:18]: This concept was called `peer` store even in the built-in libp2p deprecated version when they were supporting peer store.

**Review by Ja7ad — APPROVED** [review]

File: network/peermgr.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (b00f) [reply network/peermgr.go:18]: This should be passed from node module and be inside the `pactus` folder.

Comment (b00f) [inline network/peermgr.go:18]: This should be passed from node module and be inside the `pactus` folder.
consider add a private config for that

File: network/peermgr.go [hunk 2]
```diff
@@ -210,3 +229,35 @@ func (mgr *peerMgr) NumOutbound() int {
...
+
+	data, err := json.Marshal(ps)
+	if err != nil {
+		return err
+	}
+
+	return util.WriteFile(PeerStorePath, data)
+}
+
+func loadPeerStore() ([]lp2ppeer.AddrInfo, error) {
```
Comment (b00f) [inline network/peermgr.go:250]: Make is member of `peerMgr`

**Review by b00f — CHANGES_REQUESTED** [review]

File: network/peermgr.go (hunk 1 above)
Comment (kehiy) [reply network/peermgr.go:18]: Can we add it as a network config? or it must be node config?

File: network/peermgr.go [hunk 3]
```diff
@@ -36,27 +41,38 @@ type peerMgr struct {
 func newPeerMgr(ctx context.Context, h lp2phost.Host,
 	conf *Config, log *logger.SubLogger,
 ) *peerMgr {
-	peers := make(map[lp2ppeer.ID]*peerInfo)
+	var err error
+	peerStore := make([]lp2ppeer.AddrInfo, 0)
+	if util.PathExists(PeerStorePath) {
```
[the remark this answers is not in the document]
Comment (kehiy) [reply network/peermgr.go:46]: done

File: network/peermgr.go (hunk 1 above)
Comment (b00f) [reply network/peermgr.go:18]: It should be in `network` config. You can set it in [MakeConfig](https://github.com/pactus-project/pactus/blob/c1082588a918b14b73925c1cd95abce098154dca/cmd/cmd.go#L458) 
Feel free to pass workingDir to MakeConfig

Comment (kehiy) [reply network/peermgr.go:18]: I've set it as a private config but added the value in the default function. is that OK?

File: config/config.go [hunk 4]
```diff
@@ -204,6 +205,7 @@ func DefaultConfigLocalnet() *Config {
 	conf.Network.DefaultPort = 0
 	conf.Network.ForcePrivateNetwork = true
 	conf.Network.EnableMdns = true
+	conf.Network.PeerStorePath = "localnet-peers.json"
```
Comment (b00f) [inline config/config.go:208]: The working directory is all about one instance. 
If we want to to follow this, then we should have `testnet-config`, `testnet-wallet`, etc

File: network/errors.go [hunk 5]
```diff
@@ -38,3 +38,12 @@ type LibP2PError struct {
...
 	return fmt.Sprintf("libp2p error: %s", e.Err.Error())
 }
+
+// PeerStoreError is returned when an loading or saving permanent peer-store encounters an error.
+type PeerStoreError struct {
+	Err error
+}
+
+func (e PeerStoreError) Error() string {
+	return fmt.Sprintf("libp2p error: %s", e.Err.Error())
```
Comment (b00f) [inline network/errors.go:48]:
```suggestion
	return e.Err.Error()
```

File: network/peermgr.go [hunk 6]
```diff
@@ -210,3 +226,48 @@ func (mgr *peerMgr) NumOutbound() int {
...
+		return err
+	}
+
+	return util.WriteFile(mgr.peerStorePath, data)
+}
+
+func loadPeerStore(path string) ([]lp2ppeer.AddrInfo, error) {
+	peerStore := make([]lp2ppeer.AddrInfo, 0)
+
+	if !util.PathExists(path) {
```
Comment (b00f) [inline network/peermgr.go:250]: I think we don't need this check. Line 256 will return error

File: network/config.go [hunk 7]
```diff
@@ -48,6 +49,7 @@ func DefaultConfig() *Config {
 		ForcePrivateNetwork:  false,
 		DefaultPort:          0,
 		IsBootstrapper:       false,
+		PeerStorePath:        "peers.json",
```
Comment (b00f) [inline network/config.go:52]: Don't we need to have absolute path?

**Review by b00f — CHANGES_REQUESTED** [review]

File: network/config.go (hunk 7 above)
Comment (kehiy) [reply network/config.go:52]: No, as I tested it will locate the file on `pactus/peers.json`.

**Review by b00f — APPROVED** [review]

**Pull request merged**